### PR TITLE
Add storage migration to MemoryManager (v6 to v7)

### PR DESCRIPTION
**Pull request by przydatek**
Add functionality to migrate storage from v6 to v7, where v7 uses `ic_stable_structures::MemoryManager`. The migration does not copy the entire data, but rather uses 2nd page of the memory (which was unused/reserved so far) for `MemoryManager`'s metadata.  The anchor-memory starts in both versions from the 3rd page.
The migration is not enabled yet, and will be controlled by a flag added in a subsequent MR.
(GIX-1423)

### CODE REVIEW

File: src/internet_identity/src/storage.rs [hunk 1]
```diff
@@ -396,6 +425,79 @@ impl<M: Memory + Clone> Storage<M> {
         }
     }
 
+    #[allow(dead_code)]
+    pub fn from_memory_v6_to_v7(memory: M) -> Option<Self> {
+        let maybe_storage_v6 = Self::from_memory(memory.clone());
+        maybe_storage_v6.as_ref()?;
```
Comment (frederikrothenberger) [inline src/internet_identity/src/storage.rs:431]: What is the purpose of this line?

File: src/internet_identity/src/storage.rs [hunk 2]
```diff
@@ -396,6 +425,79 @@ impl<M: Memory + Clone> Storage<M> {
...
+        let storage_v6 = maybe_storage_v6.unwrap();
```
Comment (frederikrothenberger) [inline src/internet_identity/src/storage.rs:432]: It would be nicer to return `None` on empty rather than panicking, no?

File: src/internet_identity/src/storage.rs [hunk 3]
```diff
@@ -396,6 +425,79 @@ impl<M: Memory + Clone> Storage<M> {
...
+                std::mem::size_of::<mm::Header>(),
+            )
+        };
+        let mut mm_header_memory = RestrictedMemory::new(memory.clone(), 1..2);
+        let mut writer = Writer::new(&mut mm_header_memory, 0);
+        writer
+            .write_all(mm_header_bytes)
+            .expect("bug: failed to grow memory");
+        // Update bucket-to-memory assignments.
+        let buckets_offset: u64 = 8 + 32 + 255 * 8;
```
Comment (frederikrothenberger) [inline src/internet_identity/src/storage.rs]: Didn't you want to add some comments here?

File: src/internet_identity/src/storage.rs [hunk 4]
```diff
@@ -396,6 +425,100 @@ impl<M: Memory + Clone> Storage<M> {
...
+        // --------------------------------------------------
+        // ...
+        // --------------------------------------------------
+        // Size of memory 254 (in pages)         ↕ 8 bytes
+        // -------------------------------------------------- <- Bucket allocations
+        // ...
+        let buckets_offset: u64 = (3 + 1 + 2 + 2) + 32 + (255 * 8);
+        let mut writer = Writer::new(&mut mm_header_memory, buckets_offset);
+        let mut bucket_to_memory = [mm::UNALLOCATED_BUCKET_MARKER; mm::MAX_NUM_BUCKETS as usize];
+        for i in 0..num_allocated_buckets {
```
Comment (frederikrothenberger) [inline src/internet_identity/src/storage.rs:512]: You could avoid the loop by using `splice`:
```
let mut bucket_to_memory = vec![mm::UNALLOCATED_BUCKET_MARKER; mm::MAX_NUM_BUCKETS as usize];
bucket_to_memory.splice(
    0..num_allocated_buckets as usize,
    vec![0; num_allocated_buckets as usize],
);
```

Requires using vec though, so up to you.

File: src/internet_identity/src/storage.rs [hunk 5]
```diff
@@ -311,7 +338,9 @@ impl<M: Memory + Clone> Storage<M> {
             header_memory,
             anchor_memory,
             maybe_memory_manager,
-        }
+        };
+        storage.flush();
```
Comment (frederikrothenberger) [inline src/internet_identity/src/storage.rs:342]: Why is this flush now necessary?

File: src/internet_identity/src/storage/tests.rs [hunk 6]
```diff
@@ -80,6 +82,112 @@ fn should_recover_header_from_memory_v7() {
     assert_eq!(storage.version(), 7);
 }
 
+fn add_test_anchor_data<M: Memory + Clone>(storage: &mut Storage<M>, number_of_anchors: usize) {
+    for _ in 0..number_of_anchors {
+        let (anchor_number, mut anchor) = storage
+            .allocate_anchor()
+            .expect("Failure allocating an anchor.");
+        anchor
+            .add_device(sample_device())
```
Comment (frederikrothenberger) [inline src/internet_identity/src/storage/tests.rs:91]: It would be nice to make the devices different (i.e. depend on the index) because this would detect mix-ups in the data.

File: src/internet_identity/src/storage/tests.rs [hunk 7]
```diff
@@ -80,6 +82,112 @@ fn should_recover_header_from_memory_v7() {
...
+        storage.write(anchor_number, anchor.clone()).unwrap();
+    }
+}
+
+fn range_as_hex(memory: &RefCell<Vec<u8>>, offset: u64, length: usize) -> String {
+    let mut buf = vec![0u8; length];
+    memory.read(offset, &mut buf);
+    let s = hex::encode(buf);
+    let mut answer = String::new();
+    for (i, c) in s.chars().enumerate() {
```
Comment (frederikrothenberger) [inline src/internet_identity/src/storage/tests.rs:102]: Hm, I stumbled on this loop because it was not immediately obvious to me that you also do formatting here.
Either add a comment or drop the formatting entirely.

File: src/internet_identity/src/storage.rs (hunk 5 above)
Comment (przydatek) [reply src/internet_identity/src/storage.rs:342]: So that a new, empty storage (without any anchors) is properly initialised.

File: src/internet_identity/src/storage.rs (hunk 1 above)
Comment (przydatek) [reply src/internet_identity/src/storage.rs]: It checks whether `maybe_storage_v6` is `Some` yet without unwrapping (and returns `None` immediately if the value is `None`)

File: src/internet_identity/src/storage.rs (hunk 2 above)
Comment (przydatek) [reply src/internet_identity/src/storage.rs]: It will not panic, due to the check on the previous line.
But I noticed now that both lines can be combined, so I did that, and hopefully it is more clear now. Also, I added a test for the case when `maybe_storage_v6` is `None`.

File: src/internet_identity/src/storage.rs (hunk 4 above)
Comment (przydatek) [reply src/internet_identity/src/storage.rs:511]: Thanks for pointing it out, I'll leave as-is though, ok?

File: src/internet_identity/src/storage/tests.rs (hunk 7 above)
Comment (przydatek) [reply src/internet_identity/src/storage/tests.rs:105]: Added a comment (with the formatting it is easier to debug).

File: src/internet_identity/src/storage/tests.rs (hunk 6 above)
Comment (przydatek) [reply src/internet_identity/src/storage/tests.rs]: Good idea, done.

**Review by przydatek — COMMENTED** [review]
Thanks for the review @frederikrothenberger, PTAL.

File: src/internet_identity/src/storage.rs (hunk 1 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/storage.rs]: TIL! Thanks, I thought it only worked on `Result`.

File: src/internet_identity/src/storage.rs (hunk 5 above)
Comment (frederikrothenberger) [reply src/internet_identity/src/storage.rs:342]: So up to now, the `storage` was initialized only half-way and then `flush`ed later. Definitely a good change. 👍

**Review by frederikrothenberger — APPROVED** [review]
LGTM, thanks!
